File: Prototype/BigGp/src/Model/SmallNodes/Expressions/Logic/Childs/ComparisonExpression.java
package Model.SmallNodes.Expressions.Logic.Childs;

import Model.BaseNodes.ForStatement.ForStatement;
import Model.Interfaces.Crossable;
import Model.Interfaces.SubtreeMutable;
import Model.Node;
import Model.SmallNodes.Expressions.Logic.LogicExpression;
import Model.SmallNodes.Expressions.Standart.Expression;
import Model.SmallNodes.Operators.Logic.CompareOperator;
import Serialization.Token;

import java.util.ArrayList;

public class ComparisonExpression extends Node implements Crossable, SubtreeMutable {
    private Expression leftExpression;
    private CompareOperator operator;
    private Expression rightExpression;

    @Override
    public String toString() {
        return "(" + leftExpression.toString() + operator.toString() + rightExpression.toString() + ")";
    }

    public void setExpression(Node oldExpr, Expression expression) {
        if (leftExpression.toString().equals(oldExpr.toString()) && leftExpression.hashCode() == oldExpr.hashCode()) {
            this.leftExpression = expression;
        } else {
            this.rightExpression = expression;
        }
    }

    @Override
    public void subtreeMutate() {
        if (this.getParentNode() instanceof ForStatement forStatement) {
            forStatement.setComparisonExpression(new ComparisonExpression(forStatement, (forStatement.getNEXT_COMPARISON_EXPRESSION_CHANCE())));
        } else if (this.getParentNode() instanceof LogicExpression logicExpression) {
            logicExpression.setComparisonExpression(new ComparisonExpression(this.getParentNode()));
        }
    }

    public ArrayList<Node> getChildrenAsNodes() {
        ArrayList<Node> nodes = new ArrayList<>();
        nodes.add(this);

        nodes.addAll(leftExpression.getChildrenAsNodes());
        nodes.addAll(operator.getChildrenAsNodes());
        nodes.addAll(rightExpression.getChildrenAsNodes());

        return nodes;
    }

    public ComparisonExpression(Node parentNode) {
        super(parentNode, "ComparisonExpression", true);

        double lastNextLogicExpressionChance = ((LogicExpression) parentNode).getNEXT_LOGIC_BOOLEAN_EXPRESSION_CHANCE();
        leftExpression = new Expression(this, lastNextLogicExpressionChance);
        operator = new CompareOperator(this);
        rightExpression = new Expression(this, lastNextLogicExpressionChance);
    }

    public ComparisonExpression(Node parentNode, double NewExpressionInForComparisonChance) {
        super(parentNode, "ComparisonExpression", true);

        leftExpression = new Expression(this, NewExpressionInForComparisonChance);
        operator = new CompareOperator(this);
        rightExpression = new Expression(this, NewExpressionInForComparisonChance);
    }

    public ComparisonExpression(Node parentNode, ArrayList<Token> tokens) {
        super(parentNode, "ComparisonExpression", true);
        tokens.remove(0);

        double lastNextLogicExpressionChance = ((LogicExpression) parentNode).getNEXT_LOGIC_BOOLEAN_EXPRESSION_CHANCE();
        leftExpression = new Expression(this, lastNextLogicExpressionChance, tokens);
        operator = new CompareOperator(this, tokens);
        rightExpression = new Expression(this, lastNextLogicExpressionChance, tokens);
    }

    public ComparisonExpression(Node parentNode, double NewExpressionInForComparisonChance, ArrayList<Token> tokens) {
        super(parentNode, "ComparisonExpression", true);
        tokens.remove(0);

        leftExpression = new Expression(this, NewExpressionInForComparisonChance, tokens);
        operator = new CompareOperator(this, tokens);
        rightExpression = new Expression(this, NewExpressionInForComparisonChance, tokens);
    }
}

File: Interactive.Genetic.Programming.Environment/src/Domain/Model/Nodes.Small/Expressions.Logic/ComparisonExpression.cs
using Model.Abstract;
using Model.Interfaces;
using Model.Nodes.Big.For;
using Model.Nodes.Small.Expressions.Standard;
using Model.Nodes.Small.Operators.Logic;

namespace Model.Nodes.Small.Expressions.Logic;

public class ComparisonExpression : Node, ICrossable, ISubtreeMutable
{
    private StandardExpression Left { get; set; }
    private CompareOperator Operator { get; set; }
    private StandardExpression Right { get; set; }

    public override List<Node> ChildrenAsNodes()
    {
        var nodes = new List<Node> { this };
        nodes.AddRange(Left.ChildrenAsNodes());
        nodes.AddRange(Operator.ChildrenAsNodes());
        nodes.AddRange(Right.ChildrenAsNodes());

        return nodes;
    }

    public void SetExpression(Node oldExpression, StandardExpression newExpression)
    {
        if (Left == oldExpression)
        {
            Left = newExpression;
        }
        else
        {
            Right = newExpression;
        }
    }

    public void SubtreeMutate()
    {
        if (ParentNode is ForStatement forStatement)
        {
            forStatement.ComparisonExpression = new ComparisonExpression(forStatement, forStatement.NextComparisonExpressionChance);
        }
        else if (ParentNode is LogicExpression logicExpression)
        {
            logicExpression.ComparisonExpression = new ComparisonExpression(logicExpression);
        }
    }

    public ComparisonExpression(Node parentNode)
        : base(parentNode, "ComparisonExpression", true)
    {
        var lastNextLogicExpressionChance = ParentNode is LogicExpression logicExpression
            ? logicExpression.NextLogicBooleanExpressionChance
            : throw new InvalidOperationException("Parent node is not a LogicExpression");
        
        Left = new StandardExpression(this, lastNextLogicExpressionChance);
        Operator = new CompareOperator(this);
        Right = new StandardExpression(this, lastNextLogicExpressionChance);
    }
    
    public ComparisonExpression(Node parentNode, double newExpressionInForComparisonChance)
        : base(parentNode, "ComparisonExpression", true)
    {
        Left = new StandardExpression(this, newExpressionInForComparisonChance);
        Operator = new CompareOperator(this);
        Right = new StandardExpression(this, newExpressionInForComparisonChance);
    }
    
    public ComparisonExpression(Node parentNode, List<Token> tokens)
        : base(parentNode, "ComparisonExpression", true)
    {
        tokens.RemoveAt(0);
        
        var lastNextLogicExpressionChance = ParentNode is LogicExpression logicExpression
            ? logicExpression.NextLogicBooleanExpressionChance
            : throw new InvalidOperationException("Parent node is not a LogicExpression");
        
        Left = new StandardExpression(this, tokens, lastNextLogicExpressionChance);
        Operator = new CompareOperator(this, tokens);
        Right = new StandardExpression(this, tokens, lastNextLogicExpressionChance);
    }
    
    public ComparisonExpression(Node parentNode, List<Token> tokens, double newExpressionInForComparisonChance)
        : base(parentNode, "ComparisonExpression", true)
    {
        tokens.RemoveAt(0);
        
        Left = new StandardExpression(this, tokens, newExpressionInForComparisonChance);
        Operator = new CompareOperator(this, tokens);
        Right = new StandardExpression(this, tokens, newExpressionInForComparisonChance);
    }

    public override string ToString() => $"{Left} {Operator} {Right}";
}